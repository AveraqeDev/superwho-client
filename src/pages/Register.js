import React, { useState, useEffect } from 'react';
import { withRouter } from 'react-router-dom';
import { Input, Button } from '../components/Utils';
import { Link } from 'react-router-dom';

import UserApiService from '../services/user-api-service';

import '../styles/Register.css';

export const SidePanel = React.memo(function SidePanel() {
  return (
    <div className='SidePanel'>
      <h1>SuperWho?</h1>
      <div className="SidePanel__message">
        <h2>Hello friend!</h2>
        <p>To get started, please create a username and password!</p>
      </div>
    </div>
  );
});

export const RegisterForm = React.memo(function RegisterForm({ onRegister }) {
  const [error, setError] = useState(null);
  
  const onSubmit = e => {
    e.preventDefault();
    const { username, password, confirmPassword } = e.target;
    
    setError(null);
    if(password.value !== confirmPassword.value) {
      setError('Passwords do not match!');
      return;
    }

    UserApiService.registerUser({
      username: username.value,
      password: password.value
    })
      .then(() => {
        username.value = '';
        password.value = '';
        confirmPassword.value = '';
        onRegister();
      })
      .catch(res => {
        setError(res.error);
      })
  };
  
  return (
    <form className='RegisterForm' onSubmit={onSubmit}>
      <h2 className="Header">Sign Up</h2>
      <div role='alert'>
        {error && <p className='error'>{error}</p>}
      </div>

      <div className='RegisterForm__username-container'>
        <label htmlFor='RegisterForm__username'>
          Username
        </label>
        <Input 
          required
          name='username'
          id='RegisterForm__username'
          placeholder='Enter Your Username'
        />
      </div>
      <div className='RegisterForm__password-container'>
        <label htmlFor='RegisterForm__password'>
          Password
        </label>
        <Input 
          required
          name='password'
          type='password'
          id='RegisterForm__password'
          placeholder='Enter Your Password'
        />
      </div>
      <div className='RegisterForm__confirmPassword-container'>
        <label htmlFor='RegisterForm__confirmPassword'>
          Confirm Password
        </label>
        <Input 
          required
          name='confirmPassword'
          type='password'
          id='RegisterForm__confirmPassword'
          placeholder='Re-enter Your Password'
        />
      </div>

      <Button className='RegisterForm__button' type='submit'>
        Sign Up
      </Button>
      <p className='RegisterForm__login'>Already have an account? <Link className='RegisterForm__login-link' to='/login'>Log In</Link></p>
    </form>
  );
});

const Register = (props) => {
  const [windowWidth, setWindowWidth] = useState(0);
  const resizeWindow = () => {
    setWindowWidth(window.innerWidth);
  };

  useEffect(() => {
    resizeWindow();
    window.addEventListener("resize", resizeWindow);
    return () => window.removeEventListener("resize", resizeWindow);
  }, []);

  const onRegister = () => {
    props.history.push('/login');
  };

  if(windowWidth <= 760) {
    return (
      <div className='Register'>
        <RegisterForm onRegister={onRegister} />
      </div>
    );
  } else {
    return (
      <div className='Register'>
        <SidePanel />
        <RegisterForm onRegister={onRegister} />
      </div>
    );
  }
};

export default withRouter(Register);