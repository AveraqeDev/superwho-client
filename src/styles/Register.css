.Register {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100vw;
  height: 100vh;
}

.SidePanel {
  width: 50%;
  height: 100%;
  background: #2A9BE4;
  box-shadow: 4px 0 4px rgba(0, 0, 0, 0.25);
  color: #FFFFFF;
}

.SidePanel > h1 {
  font-size: 50px;
  padding: 2rem;
}

.SidePanel__message {
  padding: 5rem;
}

.SidePanel__message > h2 {
  margin-bottom: 0;
  font-size: 40px;
}

.SidePanel__message > p {
  font-size: 20px;
}

.RegisterForm {
  width: 50%;
  padding: 0 5rem;
  margin: auto auto;
  min-width: 415px;
}

.Header {
  font-size: 40px;
  margin-bottom: 40px;
}

.RegisterForm__username-container > label {
  font-size: 20px;
}

#RegisterForm__username {
  width: 100%;
  margin: .5rem 0;
  padding: 12px;
  color: black;
}

.RegisterForm__password-container > label {
  font-size: 20px;
}

#RegisterForm__password {
  width: 100%;
  margin: .5rem 0;
  padding: 12px;
  color: black;
}

.RegisterForm__confirmPassword-container > label {
  font-size: 20px;
}

#RegisterForm__confirmPassword {
  width: 100%;
  margin: .5rem 0;
  padding: 12px;
  color: black;
}

.RegisterForm__button {
  width: 100%;
  font-size: 20px;
  padding: 7px;
  margin-top: 2rem;
}

.RegisterForm__button:hover {
  opacity: 0.8;
}

.RegisterForm__login {
  text-align: center;
  color: #CBD0D6;
  font-weight: bold;
  font-size: 16px;
  padding: 1rem;
}

.RegisterForm__login-link {
  text-decoration: none;
  color: #2A9BE4;
}

.RegisterForm__login-link:hover {
  opacity: 0.8;
}

@media screen and (max-width: 900px) {
  .SidePanel__message {
    padding: 5rem 2.5rem;
  }
}

@media screen and (max-width: 760px) {
  .RegisterForm {
    width: 100%;
    padding: 0 5rem;
  }
}