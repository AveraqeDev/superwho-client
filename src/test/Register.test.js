import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Register, { SidePanel, RegisterForm } from '../pages/Register';

describe('Register', () => {
  describe('<SidePanel />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<SidePanel />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <SidePanel />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<RegisterForm />', () => {
    const testData = {
      onRegister: () => {}
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <RegisterForm onRegister={testData.onRegister} />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <RegisterForm onRegister={testData.onRegister} />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Register />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <Register />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <Register />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });
});