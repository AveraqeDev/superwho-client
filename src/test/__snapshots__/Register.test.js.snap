// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register <Register /> renders the UI as expected 1`] = `
<div
  className="Register"
>
  <form
    className="RegisterForm"
    onSubmit={[Function]}
  >
    <h2
      className="Header"
    >
      Sign Up
    </h2>
    <div
      role="alert"
    />
    <div
      className="RegisterForm__username-container"
    >
      <label
        htmlFor="RegisterForm__username"
      >
        Username
      </label>
      <input
        className="Input "
        id="RegisterForm__username"
        name="username"
        placeholder="Enter Your Username"
        required={true}
      />
    </div>
    <div
      className="RegisterForm__password-container"
    >
      <label
        htmlFor="RegisterForm__password"
      >
        Password
      </label>
      <input
        className="Input "
        id="RegisterForm__password"
        name="password"
        placeholder="Enter Your Password"
        required={true}
        type="password"
      />
    </div>
    <div
      className="RegisterForm__confirmPassword-container"
    >
      <label
        htmlFor="RegisterForm__confirmPassword"
      >
        Confirm Password
      </label>
      <input
        className="Input "
        id="RegisterForm__confirmPassword"
        name="confirmPassword"
        placeholder="Re-enter Your Password"
        required={true}
        type="password"
      />
    </div>
    <button
      className="Button RegisterForm__button"
      type="submit"
    >
      Sign Up
    </button>
    <p
      className="RegisterForm__login"
    >
      Already have an account? 
      <a
        className="RegisterForm__login-link"
        href="/login"
        onClick={[Function]}
      >
        Log In
      </a>
    </p>
  </form>
</div>
`;

exports[`Register <RegisterForm /> renders the UI as expected 1`] = `
<form
  className="RegisterForm"
  onSubmit={[Function]}
>
  <h2
    className="Header"
  >
    Sign Up
  </h2>
  <div
    role="alert"
  />
  <div
    className="RegisterForm__username-container"
  >
    <label
      htmlFor="RegisterForm__username"
    >
      Username
    </label>
    <input
      className="Input "
      id="RegisterForm__username"
      name="username"
      placeholder="Enter Your Username"
      required={true}
    />
  </div>
  <div
    className="RegisterForm__password-container"
  >
    <label
      htmlFor="RegisterForm__password"
    >
      Password
    </label>
    <input
      className="Input "
      id="RegisterForm__password"
      name="password"
      placeholder="Enter Your Password"
      required={true}
      type="password"
    />
  </div>
  <div
    className="RegisterForm__confirmPassword-container"
  >
    <label
      htmlFor="RegisterForm__confirmPassword"
    >
      Confirm Password
    </label>
    <input
      className="Input "
      id="RegisterForm__confirmPassword"
      name="confirmPassword"
      placeholder="Re-enter Your Password"
      required={true}
      type="password"
    />
  </div>
  <button
    className="Button RegisterForm__button"
    type="submit"
  >
    Sign Up
  </button>
  <p
    className="RegisterForm__login"
  >
    Already have an account? 
    <a
      className="RegisterForm__login-link"
      href="/login"
      onClick={[Function]}
    >
      Log In
    </a>
  </p>
</form>
`;

exports[`Register <SidePanel /> renders the UI as expected 1`] = `
<div
  className="SidePanel"
>
  <h1>
    SuperWho?
  </h1>
  <div
    className="SidePanel__message"
  >
    <h2>
      Hello friend!
    </h2>
    <p>
      To get started, please create a username and password!
    </p>
  </div>
</div>
`;
